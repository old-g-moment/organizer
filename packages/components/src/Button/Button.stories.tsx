import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import SaveIcon from '@mui/icons-material/Save';

import { Button } from './Button';

export default {
  title: 'Button',
  component: Button,
  parameters: {
    docs: {
      description: {
        component: 'Component that will be used everywhere and act as a button',
      },
    },
  },
  argTypes: {
    content: {
      description: 'Content to be displayed inside the button',
    },
    loading: {
      description: 'Indicates whether to show a loader or not',
    },
    className: {
      description: 'A custom tailwind class to be applied on an element',
    },
    variant: {
      description: 'Shows to the user a different button styles depending on the passed variant',
    },
    onClick: {
      description: 'Funtion that will be invoked when a user clicks a button',
    },
    type: {
      description: 'Specifies what kind of button is this',
    },
    style: {
      description: 'Custom css styles passed as object',
    },
    size: {
      description: 'Determines the size of the button',
    },
    startIcon: {
      description: 'Icon of the button presented on the left side of the content',
    },
    endIcon: {
      description: 'Icon of the button presented on the right side of the content',
    },
    disabled: {
      description:
        'Indicates whether the button is disabled or not and if it is disabled then the button becomes unclickable and its styles alter to simulate unavailability',
    },
  },
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />;

export const Default = Template.bind({});
Default.args = {
  content: 'Default button',
};

export const WithIcon = Template.bind({});
WithIcon.args = {
  content: <SaveIcon />,
};

export const WithCustomClass = Template.bind({});
WithCustomClass.args = {
  content: 'Custom class',
  className: 'p-10 bg-red-500',
};

export const WithLoading = Template.bind({});
WithLoading.args = {
  content: 'Loading button',
  loading: true,
};

export const WithErrorVariant = Template.bind({});
WithErrorVariant.args = {
  content: 'Error button',
  variant: 'error',
};

export const WithInfoVariant = Template.bind({});
WithInfoVariant.args = {
  content: 'Info button',
  variant: 'info',
};

export const WithWarningVariant = Template.bind({});
WithWarningVariant.args = {
  content: 'Warning button',
  variant: 'warning',
};

export const WithSuccessVariant = Template.bind({});
WithSuccessVariant.args = {
  content: 'Success button',
  variant: 'success',
};

export const WithCustomStyles = Template.bind({});
WithCustomStyles.args = {
  content: 'Custom styles',
  style: { border: '5px solid red' },
};

export const WithSize = Template.bind({});
WithSize.args = {
  content: 'With size',
  size: 'large',
};

export const WithStartIcon = Template.bind({});
WithStartIcon.args = {
  content: 'Start icon',
  startIcon: <SaveIcon />,
};

export const WithEndIcon = Template.bind({});
WithEndIcon.args = {
  content: 'End icon',
  endIcon: <SaveIcon />,
};

export const WithDisabled = Template.bind({});
WithDisabled.args = {
  content: 'Disabled button',
  disabled: true,
};

export const WithDisabledAndLoading = Template.bind({});
WithDisabledAndLoading.args = {
  content: 'Disabled loading button',
  disabled: true,
  loading: true,
};
