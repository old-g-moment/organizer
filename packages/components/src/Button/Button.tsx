import React, { ComponentProps, CSSProperties, ReactNode } from 'react';
import LoadingButton from '@mui/lab/LoadingButton';
import ButtonMui from '@mui/material/Button';
import SaveIcon from '@mui/icons-material/Save';

export type Variant = 'success' | 'warning' | 'info' | 'error' | 'default' | 'disabled';

export type Size = 'small' | 'medium' | 'large';

export const variantMap: Record<Variant, string> = {
  error: 'bg-red-600 text-slate-50',
  info: 'bg-blue-600 text-slate-50',
  warning: 'bg-orange-500 text-slate-50',
  success: 'bg-green-500 text-slate-50',
  default: 'bg-slate-700 text-slate-50',
  disabled: 'bg-gray-300 text-gray-400',
};

export type Props = {
  loading?: boolean;
  content: ReactNode;
  className?: string;
  variant?: Variant;
  onClick?: () => void;
  type?: ComponentProps<'button'>['type'];
  style?: CSSProperties;
  size?: Size;
  startIcon?: ReactNode;
  endIcon?: ReactNode;
  disabled?: boolean;
};

export const Button = ({
  content,
  className = '',
  loading = false,
  onClick,
  variant = 'default',
  type = 'button',
  style = {},
  size = 'medium',
  startIcon = null,
  endIcon = null,
  disabled = false,
}: Props) => {
  const classNames = [disabled ? variantMap.disabled : variantMap[variant], className].join(' ');

  if (loading)
    return (
      <LoadingButton
        type={type}
        variant="contained"
        className={classNames}
        style={style}
        size={size}
        loading
        loadingPosition="start"
        startIcon={<SaveIcon />}
      >
        Loading...
      </LoadingButton>
    );

  return (
    <ButtonMui
      type={type}
      variant="contained"
      className={classNames}
      style={style}
      size={size}
      disabled={disabled}
      onClick={onClick}
      startIcon={startIcon}
      endIcon={endIcon}
    >
      {content}
    </ButtonMui>
  );
};
