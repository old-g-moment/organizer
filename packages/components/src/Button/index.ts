import { Button, Props, Size, Variant, variantMap } from './Button';

export type { Props, Size, Variant, variantMap };
export default Button;
