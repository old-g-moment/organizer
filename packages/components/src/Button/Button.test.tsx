import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import SaveIcon from '@mui/icons-material/Save';

import { Button, Variant } from './Button';

const testButtonVariantToHaveClassName = (variant: Variant, className: string) => {
  render(<Button content="Custom variant button" variant={variant} />);
  const button = screen.getByRole('button', { name: /custom variant button/i });
  expect(button).toBeVisible();
  expect(button).toHaveClass(className);
};

describe('Button component tests', () => {
  it('Renders button content as text', () => {
    render(<Button content="Hello tests" />);
    expect(screen.getByRole('button', { name: /hello tests/i })).toBeVisible();
  });

  it('Renders button content as icon', () => {
    render(<Button content={<SaveIcon />} />);
    expect(screen.getByRole('button')).toBeVisible();
    expect(screen.getByTestId('SaveIcon')).toBeVisible();
  });

  it('Renders button with custom class', () => {
    render(<Button content="Custom class button" className="p-10 bg-red-500" />);
    const button = screen.getByRole('button', { name: /custom class button/i });
    expect(button).toBeVisible();
    expect(button).toHaveClass('p-10 bg-red-500');
  });

  it('Renders button with default variant', () => {
    testButtonVariantToHaveClassName('default', 'bg-slate-700');
  });

  it('Renders button with disabled variant', () => {
    testButtonVariantToHaveClassName('disabled', 'bg-gray-300');
  });

  it('Renders button with error variant', () => {
    testButtonVariantToHaveClassName('error', 'bg-red-600');
  });

  it('Renders button with info variant', () => {
    testButtonVariantToHaveClassName('info', 'bg-blue-600');
  });

  it('Renders button with warning variant', () => {
    testButtonVariantToHaveClassName('warning', 'bg-orange-500');
  });

  it('Renders button with success variant', () => {
    testButtonVariantToHaveClassName('success', 'bg-green-500');
  });

  it('Renders button with loading state', () => {
    render(<Button content="Loading button" loading />);
    expect(screen.getByRole('button', { name: /loading.../i })).toBeVisible();
    const saveIcon = screen.getByTestId('SaveIcon');
    expect(saveIcon).not.toBeVisible();
    expect(saveIcon).toBeInTheDocument();
  });

  it('Renders button with click action', async () => {
    const clickAction = jest.fn();
    render(<Button content="Click" onClick={clickAction} />);
    const button = screen.getByRole('button', { name: /click/i });
    expect(button).toBeVisible();
    await userEvent.click(button);
    expect(clickAction).toHaveBeenCalled();
  });

  it('Renders button with type', async () => {
    render(<Button content="Type" type="submit" />);
    const button = screen.getByRole('button', { name: /type/i });
    expect(button).toBeVisible();
    expect(button).toHaveAttribute('type', 'submit');
  });

  it('Renders button with custom styles', async () => {
    render(<Button content="Custom styles" style={{ border: '2px solid red' }} />);
    const button = screen.getByRole('button', { name: /custom styles/i });
    expect(button).toBeVisible();
    expect(button).toHaveStyle({ border: '2px solid red' });
  });

  it('Renders button with size', async () => {
    render(<Button content="Size" size="large" />);
    const button = screen.getByRole('button', { name: /size/i });
    expect(button).toBeVisible();
    expect(button).toHaveClass('MuiButton-sizeLarge');
  });

  it('Renders button with disabled state', async () => {
    render(<Button content="Disabled" disabled />);
    const button = screen.getByRole('button', { name: /disabled/i });
    expect(button).toBeVisible();
    expect(button).toBeDisabled();
    expect(button).toHaveClass('bg-gray-300 text-gray-400');
  });

  it('Renders button with start icon', async () => {
    const { asFragment } = render(<Button content="Start icon" startIcon={<SaveIcon />} />);
    expect(screen.getByRole('button', { name: /start icon/i })).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });

  it('Renders button with end icon', async () => {
    const { asFragment } = render(<Button content="End icon" endIcon={<SaveIcon />} />);
    expect(screen.getByRole('button', { name: /end icon/i })).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });
});
