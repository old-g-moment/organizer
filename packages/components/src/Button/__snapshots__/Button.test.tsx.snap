// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component tests Renders button with end icon 1`] = `
<DocumentFragment>
  <button
    class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root bg-slate-700 text-slate-50  css-sghohy-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    End icon
    <span
      class="MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="SaveIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
        />
      </svg>
    </span>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;

exports[`Button component tests Renders button with start icon 1`] = `
<DocumentFragment>
  <button
    class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root bg-slate-700 text-slate-50  css-sghohy-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="SaveIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
        />
      </svg>
    </span>
    Start icon
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;
